boa_engine/src/syntax/ast/expression/optional.rs

```diff
@@ -0,0 +1,195 @@
+use boa_interner::{Interner, Sym, ToInternedString};
+
+use crate::syntax::ast::{join_nodes, ContainsSymbol};
+
+use super::{access::PropertyAccessField, Expression};
+
+/// List of valid expressions in an [`Optional`] chain.
+#[cfg_attr(feature = "deser", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq)]
+pub enum OptionalItemKind {
+    /// A property access (`a?.prop`).
+    SimplePropertyAccess {
+        /// The field accessed.
+        field: PropertyAccessField,
+    },
+    /// A private property access (`a?.#prop`).
+    PrivatePropertyAccess {
+        /// The private property accessed.
+        field: Sym,
+    },
+    /// A function call (`a?.(arg)`).
+    Call {
+        /// The args passed to the function call.
+        args: Box<[Expression]>,
+    },
+}
+
+impl OptionalItemKind {
+    #[inline]
+    pub(crate) fn contains_arguments(&self) -> bool {
+        match self {
+            OptionalItemKind::SimplePropertyAccess { field } => field.contains_arguments(),
+            OptionalItemKind::PrivatePropertyAccess { .. } => false,
+            OptionalItemKind::Call { args } => args.iter().any(Expression::contains_arguments),
+        }
+    }
+    #[inline]
+    pub(crate) fn contains(&self, symbol: ContainsSymbol) -> bool {
+        match self {
+            OptionalItemKind::SimplePropertyAccess { field } => field.contains(symbol),
+            OptionalItemKind::PrivatePropertyAccess { .. } => false,
+            OptionalItemKind::Call { args } => args.iter().any(|e| e.contains(symbol)),
+        }
+    }
+}
+
+/// Item within an [`Optional`] chain.
+///
+/// An item within an `Optional` chain can be either shorted or non-shorted. A shorted item
+/// (`?.item`) will force the expression to return `undefined` if the target is `undefined` or `null`.
+/// In contrast, a non-shorted item (`.item`) will try to access the property, even if the target
+/// is `undefined` or `null`.
+#[cfg_attr(feature = "deser", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Clone, Debug, PartialEq)]
+pub struct OptionalItem {
+    kind: OptionalItemKind,
+    shorted: bool,
+}
+
+impl OptionalItem {
+    /// Creates a new `OptionalItem`.
+    pub fn new(kind: OptionalItemKind, shorted: bool) -> Self {
+        Self { kind, shorted }
+    }
+    /// Gets the kind of optional chain item.
+    pub fn kind(&self) -> &OptionalItemKind {
+        &self.kind
+    }
+
+    /// Returns `true` if the item short-circuits the [`Optional`] chain when `undefined`/`null`.
+    pub fn shorted(&self) -> bool {
+        self.shorted
+    }
+
+    #[inline]
+    pub(crate) fn contains_arguments(&self) -> bool {
+        self.kind.contains_arguments()
+    }
+
+    #[inline]
+    pub(crate) fn contains(&self, symbol: ContainsSymbol) -> bool {
+        self.kind.contains(symbol)
+    }
+}
+
+impl ToInternedString for OptionalItem {
+    #[inline]
```

Comment:
I would not use `#[inline]` here, might be too big of a function to inline, and probably the compiler can decide.